orchagent/vrforch.cpp

```diff
@@ -0,0 +1,174 @@
+            {
+                erase_from_queue = delOperation(request);
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Wrong operation. Check RequestParser: %s", op.c_str());
+            }
+        }
+        catch (const std::invalid_argument& e)
+        {
+            SWSS_LOG_ERROR("Parse error: %s", e.what());
+        }
+        catch (const std::logic_error& e)
+        {
+            SWSS_LOG_ERROR("Logic error: %s", e.what());
+        }
+        catch (const std::exception& e)
+        {
+            SWSS_LOG_ERROR("Exception was catched in the request parser: %s", e.what());
+        }
+        catch (...)
+        {
+            SWSS_LOG_ERROR("Unknown exception was catched in the request parser");
+        }
+        request.clear();
+
+        if (erase_from_queue)
+        {
+            it = consumer.m_toSync.erase(it);
+        }
+        else
+        {
+            ++it;
+        }
+    }
+}
+
+
+bool VRFOrch::addOperation(const VRFRequest& request)
+{
+    SWSS_LOG_ENTER();
+
+    sai_attribute_t attr;
+    vector<sai_attribute_t> attrs;
+
+    const std::string& vrf_name = request.getKeyString(0);
+    if (vrf_table_.find(vrf_name) != std::end(vrf_table_))
+    {
+        SWSS_LOG_ERROR("VRF '%s' already exists", vrf_name.c_str());
+        return true;
+    }
+
+    for (const auto& name: request.getAttrFieldNames())
+    {
+        if (name == "v4")
+        {
+            attr.id = SAI_VIRTUAL_ROUTER_ATTR_ADMIN_V4_STATE;
+            attr.value.booldata = request.getAttrBool("v4");
+        }
+        else if (name == "v6")
+        {
+            attr.id = SAI_VIRTUAL_ROUTER_ATTR_ADMIN_V6_STATE;
+            attr.value.booldata = request.getAttrBool("v6");
+        }
+        else if (name == "src_mac")
+        {
+            const auto& mac = request.getAttrMacAddress("src_mac");
+            attr.id = SAI_VIRTUAL_ROUTER_ATTR_SRC_MAC_ADDRESS;
+            memcpy(attr.value.mac, mac.getMac(), sizeof(sai_mac_t));
+        }
+        else if (name == "ttl_action")
+        {
+            attr.id = SAI_VIRTUAL_ROUTER_ATTR_VIOLATION_TTL1_PACKET_ACTION;
+            attr.value.s32 = request.getAttrPacketAction("ttl_action");
+        }
+        else if (name == "ip_opt_action")
+        {
+            attr.id = SAI_VIRTUAL_ROUTER_ATTR_VIOLATION_IP_OPTIONS_PACKET_ACTION;
+            attr.value.s32 = request.getAttrPacketAction("ip_opt_action");
+        }
+        else if (name == "l3_mc_action")
+        {
+            attr.id = SAI_VIRTUAL_ROUTER_ATTR_UNKNOWN_L3_MULTICAST_PACKET_ACTION;
+            attr.value.s32 = request.getAttrPacketAction("l3_mc_action");
+        }
+        else
+        {
+            SWSS_LOG_ERROR("Logic error: Unknown attribute: %s", name.c_str());
+            continue;
+        }
+        attrs.push_back(attr);
+    }
+
+    sai_object_id_t router_id;
+    sai_status_t status = sai_virtual_router_api->create_virtual_router(&router_id,
```

Comment:
>create_virtual_router [](start = 50, length = 21)

if the vrf exists, then we need to do a set operation, i think.

Reply:
You mean 'update' operation? Don't create a new vrf, but update existing one?

Reply:
yes

Reply:
ok. In this case 'update' operation might have two semantics:
1. remove old attributes and add the new one from the request. Currently impossible in SAI (no function to remove attributes or setting them into default value).
2. merge old attributes with new ones from the request..

I'll implement the second approach.